system/jlib/jthread.cpp

```diff
@@ -702,22 +702,45 @@ void CAsyncFor::For(unsigned num,unsigned maxatonce,bool abortFollowingException
                 return 0;
             }
         };
-        if (maxatonce==0)
+        if (maxatonce==0 || maxatonce > num)
             maxatonce = num;
-        for (i=0;(i<num)&&(i<maxatonce);i++)
-            ready.signal();
-        IArrayOf<Thread> started;
-        started.ensure(num);
-        for (i=0;i<num;i++) {
-            ready.wait();
-            if (abortFollowingException && e) break;
-            Owned<Thread> thread = new cdothread(this,shuffled?shuffler->lookup(i):i,ready,&errmutex,e);
-            thread->start();
-            started.append(*thread.getClear());
+        if (maxatonce < num)
+        {
+            for (i=0;(i<num)&&(i<maxatonce);i++)
+                ready.signal();
+            IArrayOf<Thread> started;
+            started.ensure(num);
+            for (i=0;i<num;i++) {
+                ready.wait();
+                if (abortFollowingException && e) break;
+                Owned<Thread> thread = new cdothread(this,shuffled?shuffler->lookup(i):i,ready,&errmutex,e);
+                thread->start();
+                started.append(*thread.getClear());
+            }
+            ForEachItemIn(idx, started)
+            {
+                started.item(idx).join();
+            }
         }
-        ForEachItemIn(idx, started)
+        else
         {
-            started.item(idx).join();
+            // Common case of execute all at once can be optimized a little
+            // Note that shuffle and abortFollowingException are meaningless when executing all at once
+            ready.reinit(num-1);
+            IArrayOf<Thread> started;
+            started.ensure(num);
+            for (i=0;i<num-1;i++)
+            {
+                Owned<Thread> thread = new cdothread(this,i,ready,&errmutex,e);
+                thread->start();
+                started.append(*thread.getClear());
+            }
+            Do(num-1);
+            ready.wait(); // Waiting on a single semaphore should help reduce context switches, even though we still need to join below
```

Comment:
This isn't doing what you hope for - the reinit will have set it to a number > 0, so the wait will return immediately.

You cannot supply a -ve value to sem_init, so I'm not sure you can do what you are aiming for.